File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/PatchTitleString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class PatchTitleString {
    private static final String DNCString = "?DNC?";

    public String title;

    public PatchTitleString(String title) {
        this.title = title;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static PatchTitleString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 1, "PatchTitle");
        return new PatchTitleString(parts.get(0));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof PatchTitleString)) return false;
        PatchTitleString that = (PatchTitleString) o;
        return (DNCString.equals(title) || DNCString.equals(that.title) || Objects.equals(title, that.title));
    }

    @Override
    public int hashCode() {
        return Objects.hash(title);
    }

    @Override
    public String toString() {
        return Tokens.token(title);
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/PatchTitleString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class PatchTitleString
    {
        public string title;

        public PatchTitleString(string title)
        {
            this.title = title;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static PatchTitleString FromText(string text)
        {
            var parts = Tokens.Require(text, 1, "PatchTitle");
            return new PatchTitleString(parts[0]);
        }

        public PatchTitleTyped ToPatchTitleTyped()
        {
            return new PatchTitleTyped(
                this.title
            );
        }

        public override string ToString()
        {
            return Tokens.Token(title);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not PatchTitleString other) return false;
            return (DNCString == this.title || DNCString == other.title || object.Equals(this.title, other.title));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.title);
            return h.ToHashCode();
        }
    }
}
